torch/distributions/transformed_distribution.py

```diff
@@ -181,9 +181,9 @@ def icdf(self, value):
         transform(s) and computing the score of the base distribution.
         """
         value = self._monotonize_cdf(value)
-        if self._validate_args:
-            self.base_dist._validate_sample(value)
         value = self.base_dist.icdf(value)
         for transform in self.transforms:
             value = transform(value)
+        if self._validate_args:
```

Comment:
Now that I'm taking a fresh look at this, I think it might be better to remove this check entirely. For a transformed distribution, the support will be more constrained than what is deduced through the `@support` property. For instance, for the example above, I think this will only check that all values are positive not that all values should be between `[e^lower, e^upper]`. We also do not validate samples in other distribution's `icdf`, so if we wanted to do this, we can instead move this check to the base distribution's `icdf` instead which will cover the intent of this check.

Reply:
Hi Neeraj, thanks for the review.

Would you recommend that I outright remove the validation in this `icdf` function? That's an easy enough change and will satisfy the unit test that I added. 

One question, you say, 

"We also do not validate samples in other distribution's icdf, so if we wanted to do this, we can instead move this check to the base distribution's icdf instead."

Where exactly would the sample validation logic go? Into the base distribution class?

Reply:
> Where exactly would the sample validation logic go? Into the base distribution class?

That's right. I think just checking that the `icdf` method for the base distribution yields a value within the support of the base distribution is sufficient and can be done for each of the distributions separately in their `icdf` methods. If this passes, it will pass the check for the `TransformedDistribution.support` after the transforms are applied. Checking for the latter directly isn't super useful, since `TransformedDistribution.support` check is currently too lax. e.g. for `TransformedDistribution(Uniform(5, 10), ExpTransform())`, the support will be `constraints.real` instead of the interval `[e**5, inf)`, so just directly checking that `Uniform.icdf` produces a value that is in `[5, 10]` would make more sense (not proposing that we need to do this in this PR).

Reply:
Hi Neeraj, I hope your week is off to a good start.

I am not an expert at Python, but I wanted to avoid re-implementing a bunch of logic in distribution subclasses everywhere, so I have attempted to make use of the `__init_subclass__` hook to automatically decorate subclasser's implementations of icdf with sample validation logic.

I ran into some mypy issues. It seems like mypy dislikes the `super().__init_subclass__` call and I'm not sure why. This exact issue was closed [here](https://github.com/python/mypy/issues/4660) but it seems to still occur for me. I was able to remove the `*args` and `**kwargs` params but not sure if that's the best.

Let me know how this looks!

Reply:
Looks like flake is complaining about my use of `setattr` ☹️

Reply:
> so I have attempted to make use of the __init_subclass__ hook to automatically decorate subclasser's implementations of icdf with sample validation logic.

Interesting, I have not used `__init_subclass__` earlier, so trying to get acquainted with this. I'll take a look at mypy, but my bigger concern would be not breaking backward compatibility for user defined distributions which may not have `.support` defined (I think this should be fine and it will merely raise a warning, but I'll check again).

Reply:
A case in point with regard to subclasses is behavior for `TransformedDistribution`. We would like to omit this check for `TransformedDistribution` (since the base classes are already checked) but IIUC this will instead run the check on the transformed distribution as well as the base class. In general, I think given that subclassing `Distribution` is very common, it is best to delegate this to the subclasses themselves. For this PR, I think we can simply remove this check.

Reply:
I _think_ the `__init_subclass__` usage there would have only done the check on the base class if the derived class called `super().icdf( ... )` but that could be wrong. Regardless, I don't fully understand it and I agree we should strive to not break user defined distributions, so I've removed all that.

Now the PR just adds a unit test and removes the validation call that caused the error in the first place.

Thanks!